django/db/migrations/migration.py

```diff
@@ -139,18 +175,37 @@ def unapply(self, project_state, schema_editor, collect_sql=False):
         """
         # Construct all the intermediate states we need for a reverse migration
         to_run = []
-        new_state = project_state
+
         # Phase 1
-        for operation in self.operations:
-            # If it's irreversible, error out
-            if not operation.reversible:
-                raise IrreversibleError("Operation %s in %s is not reversible" % (operation, self))
-            # Preserve new state from previous run to not tamper the same state
-            # over all operations
-            new_state = new_state.clone()
-            old_state = new_state.clone()
-            operation.state_forwards(self.app_label, new_state)
-            to_run.insert(0, (operation, old_state, new_state))
+        def _collect_operations(operations, to_state):
+            if not operations:
+                return
+
+            for operation in operations:
+                # If it's irreversible, error out
+                if not operation.reversible:
+                    raise IrreversibleError("Operation %s in %s is not reversible" % (operation, self))
+                # Preserve new state from previous run to not tamper the same state
+                # over all operations
+                to_state = to_state.clone()
+                from_state = to_state.clone()
+                operation.state_forwards(self.app_label, to_state)
+                to_run.insert(0, (operation, from_state, to_state))
+
+                # Insert injected operations from post_operation signal receivers.
+                injected_operations = emit_post_operation_signal(
+                    migration=self,
+                    operation=operation,
+                    from_state=from_state.clone(),
+                    to_state=to_state.clone(),
```

Comment:
@arthurio I think doing the clone here makes sense. You could have conditionally done it in `emit_post_operation_signal()` only if there was some `post_operation` receivers to avoid the `clone()` if it wasn't the case but given we're planning to add a few receivers for `auth` and `contenttypes` and at least one of them is going to be installed in 99% of projects I don't think it's worth doing the optimization.